_posts/2022-09-14-utilizing-state-machines-in-front-end-development.md

```diff
@@ -0,0 +1,543 @@
+    event.preventDefault();
+
+    send("SEARCH");
+
+    fetchResults(state.context.keyword)
+      .then((results) => {
+        send({ type: "SUCCESS", data: results });
+      })
+      .catch((error) => send("ERROR"));
+  };
+
+  return (
+    <div>
+      <form onSubmit={search}>
+        <input
+          onChange={(event) =>
+            send({ type: "SET_KEYWORD", data: event.target.value })
+          }
+        />
+        <button type="submit" disabled={state.value === "loading"}>
+          Search
+        </button>
+      </form>
+
+      {state.value === "initial" && <p>Hi! Enter keyword please.</p>}
+      {state.value === "loading" && <p>Searching...</p>}
+      {state.value === "failure" && <p>Oh no!</p>}
+      {state.value === "noResults" && <p>Sorry! No results.</p>}
+      {state.value === "showResults" &&
+        state.context.results.map((result) => <p>{result}</p>)}
+    </div>
+  );
+};
+```
+
+Cool, now the component's whole state is handled by XState. We have successfully moved the business logic out from the view layer handled by React. We could even move the fetching to be a part of the machine (maybe a reusable sub-machine?).
+
+And by the way, our machine looks like this now:
+
+![A visual presentation of a state machine representing a search form on a web app](/img/utilizing-state-machines-in-front-end-development/SearchFormWithXStateV2.png)
+
+The `SET_KEYWORD` event assigns the data but doesn't trigger a transition while `SUCCESS` assigns and transitions at the same time.
+
+# Conclusion
+
+State machines and statecharts are super effective tools when used correctly. I especially like the fact that when using them I'm forced to think about the behavior of the app and describe it before implementing it.
+
+State machines add formalism and make code more robust by making sure there are no impossible states. They can separate business logic from the view layer and are a great way to make reusable logic. They are easy to represent visually, which can drastically help with communicating ideas, for example.
+
+On the other hand, state machines require you to think in a new mental model and can sometimes be a little overwhelming. A state machine with hundreds of lines of config code can seem like you are just moving your spaghetti from one plate to another. And you very well might be doing just that, but I still find organizing logic is much easier with state machines.
+
+My recommendation: **add state machines to your developer toolbox!** Next time you are adding a third boolean flag to your component, think to yourself, "could this be a state machine instead?". I'll promise, soon you'll start to see state machines everywhere you look. You don't need to use any library at all, you don't even have to do a full-fledged FSM. Just replace multiple booleans with one state variable with multiple, predefined values and enjoy better quality code with fewer corner cases. If you decide to use a library, XState gets my vote.
```

Comment:
It looks like Redux and Flux with FSAs and dispatch. I'm just wondering is there a difference and how should I pick my state machine lib?

Reply:
I'll add a comment about redux to the pros section